gradle/generation/snowball.gradle

```diff
@@ -0,0 +1,118 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+apply plugin: "de.undercouch.download"
+
+configure(rootProject) {
+  ext {
+    snowballWorkDir    = file("${buildDir}/snowball")
+    snowballStemmerDir = file("${buildDir}/snowball/stemmers")
+    snowballPatchedDir = file("${buildDir}/snowball/patched")
+    snowballWebsiteDir = file("${buildDir}/snowball/website")
+    snowballDataDir    = file("${buildDir}/snowball/data")
+  }
+
+  task snowball()  {
+    description "Regenerate snowball-based sources, stopwords, and tests for ...lucene/analysis."
+    group "generation"
+
+    dependsOn ":lucene:analysis:common:snowballGen"
+  }
+
+  task downloadSnowballStemmers(type: Download) {
+    def stemmerZip = file("${snowballWorkDir}/stemmers.zip")
+
+    src "https://github.com/snowballstem/snowball/archive/53739a805cfa6c77ff8496dc711dc1c106d987c1.zip"
+    dest stemmerZip
+    onlyIfModified true
+    useETag "all"
+    cachedETagsFile file("${snowballWorkDir}/stemmers.json")
+
+    doLast {
+      ant.unzip(src: stemmerZip, dest: snowballStemmerDir, overwrite: "true") {
+        ant.cutdirsmapper(dirs: "1")
+      }
+    }
+  }
+
+  task downloadSnowballWebsite(type: Download) {
+    def websiteZip = file("${snowballWorkDir}/website.zip")
+
+    src "https://github.com/snowballstem/snowball-website/archive/ff891e74f08e7315523ee3c0cad55bb1b7831b9d.zip"
+    dest websiteZip
+    onlyIfModified true
+    useETag "all"
+    cachedETagsFile file("${snowballWorkDir}/website.json")
+
+    doLast {
+      ant.unzip(src: websiteZip, dest: snowballWebsiteDir, overwrite: "true") {
+        ant.cutdirsmapper(dirs: "1")
+      }
+    }
+  }
+
+  task downloadSnowballData(type: Download) {
+    def dataZip = file("${snowballWorkDir}/data.zip")
+
+    src "https://github.com/snowballstem/snowball-data/archive/9145f8732ec952c8a3d1066be251da198a8bc792.zip"
+    dest dataZip
+    onlyIfModified true
+    useETag "all"
+    cachedETagsFile file("${snowballWorkDir}/data.json")
+
+    doLast {
+      ant.unzip(src: dataZip, dest: snowballDataDir, overwrite: "true") {
+        ant.cutdirsmapper(dirs: "1")
+      }
+    }
+  }
+
+  task cleanSnowballCheckout(type: Delete) {
+    dependsOn downloadSnowballStemmers
+    delete snowballPatchedDir
+  }
+
+  task patchSnowball(type: Copy) {
+    dependsOn cleanSnowballCheckout
+
+    from fileTree(snowballStemmerDir) {
+      include '**/*'
+    }
+    into snowballPatchedDir
+
+    doLast {
+      ant.patch(patchfile: rootProject.file("gradle/generation/snowball.patch"), dir: snowballPatchedDir, strip: "1")
```

Comment:
ant.patch just redirects to local filesystem patch command (doesn't work on Windows). Sigh. Fine with me though.

Reply:
I think if you have patch.exe on windows then this part will work?

Reply:
Not many people do (I don't). You'd need to run under cygwin or have msys tools installed (which are recompiled for Windows). I don't think this is a problem. I don't think there is a single Windows developer who isn't fluent on other environments. If a regeneration task does not work on Windows so be it (I'd make it complain gracefully but this can come later).

Reply:
I looked into platform-independent patch command implementation but I didn't find any (jgit's is inherently broken; @kojisekig wrote something for jenkins but I didn't look into that - maybe he'll have some more insight).

Reply:
Maybe we can focus on trying to remove the patch completely (try to package up some PRs for snowball team) and it gets fixed even before their Makefile supports windows :)

I guess I optimistically hope the patch solution is a temporary one. If you look deeper into shell script and snowball GNUMakefile, it may be obvious that we have other issues to fix for windows first.

Reply:
That download task is broken in that it still executes actions even if the artifact is up to date... You can just delete the unpacked stuff and not the downloaded zip; then it'll wipe and unzip but not download again?

Reply:
@dweiss I can assure you it does not download again. Please look at the PR, maybe the names are confusing and we can clean it up? I think logic is correct.

Here is what happens (first run):
* download `github.com/snowball/xyz.zip` to `snowball.zip`. It extracts this to `stemmers/`. It saves Etag to `stemmers.json`.
* create a clean copy of `stemmers/` named `patched/`.
* apply patch to `patched/`.

Second run:
* Tries to download, but sends request with `If-None-Match: a4gk4k4...`. Remote server returns a HTTP 304 (Not Modified). This would not be the case if, e.g. someone bumped the snowball commit hash locally.
* create a clean copy of `stemmers/` named `patched/`.
* apply patch to `patched/`.

Reply:
I know it doesn't download twice but if it declared inputs/ outputs or upToDateWhen correctly then it shouldn't run any actions attached to the task (doFirst/ doLast) and it does it now, regardless of whether it downloaded the artifact or not, right (that was my impression from other sections of the code, no this patch). 

Anyway, if it's working then fine. As for different versions of patch -- throw git/ jgit's implementation in the mix... they all seem to be behaving differently in corner cases.

Reply:
note this same trick (HTTP cache revalidation) cannot be used for moman, because its server disables caching completely and doesn't send such headers. In fact the server does not even respond correctly to requests with `curl` (unless you modify User-Agent header) so it is clear they are not excited about lots of automated requests, so it should be improved some other way...

Reply:
@dweiss ok, I will look at inputs/outputs/upToDateWhen. Maybe a simpler way is for these tools to download to a file with commit's hash in the name and declare such stuff? We'd be able to fix moman to not download over and over too (cache headers cannot work over there).

Reply:
This isn't something you have to look into, Rob. I'm just adding some comments, that's it. The patch is fine.

The simplest way of making the task not download the file again would be to declare an inputs.file(path-to-zip) I think. But it's really a marginal issue, I wouldn't spend any special time on this if it works.

Reply:
@dweiss So we can add these dependencies, but I just don't want it to patch twice :) That's kinda why their is an explicit "clean" here to create patched/ work area.

I think behavior is fast enough for a first go and we can followup? It takes 26s the first time and 4s the second time for me.

I am worried about making it too smart and declaring such dependencies incorrectly. I want to make sure it does the right thing if you e.g. update commit hash/patch file, or change script, so that its easier to upgrade snowball.

Reply:
It wasn't my intention to change the patch - these were my random musings about gradle, the download task, etc. It's definitely fine for committing in and can be improved later (or never if we can convince people from snowball to generate better code).

Reply:
Yes but if you teach a man to fish... I will look into it. Its just tricky, for example patching task in my mind depends on "cleanCheckoutOfCurrentlySpecifiedCommitHash + input patch file itself". and maybe we must "touch" an output file indicating patch was applied. This way it doesn't try to apply it twice. At least this is how i would do it with something like `make`

Reply:
Declare inputs on the zip and move applying the patch to inside the download task action (you don't even have to copy the unpacked folder then), execute patching in doLast; if it fails the entire task won't be up-to-date (will be re-run)? Just a thought.